### ISSUE TITLE
Scaper and random seeds

### ISSUE THREAD

**Issue by pseeth** [justinsalamon/scaper#36]
Currently, there doesn't appear to be a way to seed scaper with something like random.seed(0) so that it produces the same mixtures given the same random seed and set of source files. Just starting this issue to discuss what changes would need to be done to accomplish it.

**Comment by justinsalamon** [comment]
Thanks for opening this issue @pseeth, this would be a great feature to add to scaper I think.

Perhaps the first issue to address is that right now scaper uses a mix of python's native `random` ([here](https://github.com/justinsalamon/scaper/blob/master/scaper/core.py#L26)) and `scipy` ([here](https://github.com/justinsalamon/scaper/blob/master/scaper/util.py#L146)) to generate/sample from the supported distributions, which is a little clunky, and probably makes seeding complicated.

So the first step would be to consistently use the same module for random number generation throughout scaper (probably scipy?). One would also have to go over the code to ensure there aren't  other places where random numbers are being generated in addition to the two places I've highlighted.

Other than that, I wonder whether it'd make sense to document the seed anywhere in scaper's output? For example it could be added to the jams annotation under `annotation.sandbox.scaper['random_seed']`, though I'm not sure whether/how useful that would be given that you might generate a whole batch of soundscapes in a single run, and so without knowing the index of the soundscape in the batch the seed wouldn't help to reproduce it.

Thoughts?

**Comment by justinsalamon** [comment]
yes, I guess we could just seed all the RNG's. 

BUT... I ran a quick experiment seeding all 3 (random, numpy, scipy), and there are a few things to note:
1. Using the same seed, `random` results in a different sequence compared to numpy and scipy (which give the same sequence) 
2. Numpy and scipy seem to share the same RNG internally, so calling either advances the internal sequence. This means it *might* be sufficient to seed just the one.
3. Whatever we do we'll need some strong tests to ensure batch generation is reproducible.

Another question is where to seed. We could seed the module itself `scaper.seed()`, or you could seed a generator when it's created `sc = scaper.Scaper(seed=123)`. Thoughts?

**Comment by pseeth** [comment]
1. Yeah that makes sense, I suppose. Gotta seed both.
2. I think they do share the same RNG. First, all google searches for "scipy random" redirect to np.random. And it seems scipy.stats at least calls into  np.random according to this [post](https://stackoverflow.com/questions/16016959/scipy-stats-seed).
3. Yes, definitely will need strong testing. Also if someone adds a source of randomness down the line that is not seeded in scaper.manual_seed() (or wherever the function ends up), this test should fail. Which means the test will need to cover all possible sources of randomness when generating soundscapes.

We could do either I think. When I use scaper, I parallelize my generation process across all my CPUs, which involves me creating a generator for each process (actually I just realized that manual seeding would break my code as it stands, generating the same soundscape over and over - but that's a user issue that I can fix in my script). So for reproducibility, I'd have to make sure to seed every generator in the second case. Maybe there's a better way to implement parallelism than what I'm doing, also. Seeding the module itself would be straightforward, like how PyTorch does it.

**Comment by justinsalamon** [comment]
Seeding with parallel processes... there be dragons. Worth thinking about carefully I guess. @bmcfee curious to hear your 2c on this - did you have to deal with this in librosa?

**Comment by pseeth** [comment]
Hm one paradigm for reproducibility with parallel processes that could work is to create all the soundscape jams files in sequence with the random seed set. That should be quick. And then you can use generate_from_jams, which should be deterministic, to parallelize the actual creation of all the audio to make it more efficient.

**Comment by bmcfee** [comment]
> @bmcfee curious to hear your 2c on this - did you have to deal with this in librosa?

Not in librosa, but it does come up in pescador and muda.

We especially had to jump through some hoops in pescador to make it work nicely across process boundaries, see [this thread](https://github.com/pescadores/pescador/pull/113).  Basically we had to ensure that RandomState objects are deep-copied at the right times (see [here](https://github.com/pescadores/pescador/blob/master/pescador/mux.py#L419-L437)).

In muda, I did something like what @pseeth suggests, separating out randomness (state generation) from rendering, so that the latter can operate entirely on deterministic state information.  I like this solution in general, especially if you can do the pre-generation with one RS object and then have deterministic synthesis sharded out across processes.  If you try to shard out the randomness across processes, you'll run into issues:

1. common access to a shared state object (if it even works) will be sensitive to execution order and the number of processes.  This is bad for reproducibility.
2. if you duplicate randomstate objects across processes, you could end up with a huge reduction of entropy.  Imagine that each process gets its own private copy of a seeded RNG, and the parameters are all otherwise the same.  Then all processes will generate the same data in the same order.  Not what you want.

**Comment by justinsalamon** [comment]
Cheers @bmcfee 

Sounds like the best option for parallelism is, as you both suggest, to pre-generate the jams objects using a single process, and then parallelize the audio instantiation. 

This is already perfectly possible in the current version, it's just up to the user to design their code that way (i.e. first generate jams in a single process, then parallelize instantiation). My question is thus - do you think there's any need to provide this functionality as part of scaper (something like `Scaper.generate_parallel()`), over leaving it to the user to manage?

**Comment by justinsalamon** [comment]
@pseeth wanna open up a PR to implement seeding?

**Comment by pseeth** [comment]
I think I could take a crack at it, but won't have time till after Aug 22 (out of office for a bit). Planning on some scaper stuff then!

**Comment by justinsalamon** [comment]
OK cool, I'll let ya know if I find myself getting to it before then, otherwise we can work on it once you're back in the office. Enjoy your down time!

**Comment by FunkMonkey** [comment]
Does it make sense to provide a RNG instead of just a seed in terms of flexibility? Could also be just a simple RNG function that returns a value between 0 and 1. ..

**Comment by bmcfee** [comment]
> Does it make sense to provide a RNG instead of just a seed in terms of flexibility? Could also be just a simple RNG function that returns a value between 0 and 1...

I usually follow sklearn's example for this, where they can accept either a seed (integer) or an existing `np.random.RandomState` object.  If the former, a randomstate object is constructed from the seed.  If the latter, it is used directly.  This makes it easy to serialize, but also share state when necessary.

**Comment by justinsalamon** [comment]
Thanks everyone for pitching in!

> I usually follow sklearn's example for this, where they can accept either a seed (integer) or an existing np.random.RandomState object. If the former, a randomstate object is constructed from the seed. If the latter, it is used directly. This makes it easy to serialize, but also share state when necessary.

I like this solution. So basically we'd add a global function a la `scaper.random_state()` which takes either a `np.random.RandomState` object or an integer (which is used to construct a RandomState object).

What should the default behavior be? i.e., if the user does not call `scaper.random_state()`, should scaper always default to the same initial state, or should it use a random initial state (e.g by making it a function of `time.time()` or something like that)? I guess the former would make more sense?

**Comment by bmcfee** [comment]
> What should the default behavior be?

Default should be to use `np.random`; that way a user can set a global random state via numpy and have it behave as expected.

**Comment by justinsalamon** [comment]
OK, after a quick offline chat with @bmcfee our current thinking is there shouldn't be a module-level RNG (instead default to using numpy.random), but there should be a Scaper (class) level RNG. So it would work like this:

- scaper (the module) does *not* have a `random_state()` function nor a global RNG
- Scaper (the class) take an RNG (or an `int` for constructing an RNG object) in it’s `__init__`. If `None` is provided, it defaults to `self.rng = numpy.random`.
- Every function that uses randomness (which are all class methods) would call `self.rng`

@pseeth sounds good?

**Comment by pseeth** [comment]
So would a refactor be needed to get rid of references to RNGs that are not the numpy one? Or am I misunderstanding?

**Comment by justinsalamon** [comment]
Yes, but not just replacing them, rather, a restructure as follows:

1. Add a `random_state` kwarg to the `Scaper` class's `init` function

The idea would be to adopt the scikit-learn API:
```
random_state : int, RandomState instance or None, optional (default=None)

If int, random_state is the seed used by the random number generator; If RandomState instance, random_state is the random number generator; If None, the random number generator is the RandomState instance used by np.random.
```
The RandomState object (constructed using provided int, or directly passed by user, or `np.random` if `None`, would be stored in a `self.random_state` class's field.

2. Pass `self.random_state` to all class methods that use a random number generator. Right now these methods make use of a global [SUPPORTED_DIST](https://github.com/justinsalamon/scaper/blob/master/scaper/core.py#L26-L30) dictionary, so that'll have to be restructured to take and use the RandomState object.

Does that sound reasonable?

**Comment by pseeth** [comment]
Ah okay, yeah it makes sense. So should all the methods in SUPPORTED_DIST be adjusted to be lambda functions taking in a RandomState and an input? That seems a bit messy. Also, what do we do with _trunc_norm? That distribution isn't listed in numpy RandomState.

**Comment by justinsalamon** [comment]
> Ah okay, yeah it makes sense. So should all the methods in SUPPORTED_DIST be adjusted to be lambda functions taking in a RandomState and an input? That seems a bit messy.

I'm open to ideas! Currently sampling always happens via a call to [`_get_value_from_dist(dist_tuple)`](https://github.com/justinsalamon/scaper/blob/master/scaper/core.py#L246-L274) which is defined globally. You could argue that since sampling a RV is not something specific to the scaper object that'd it'd make sense to keep it global. On the flip side, if we're now making the RNG be a field of the Scaper object, perhaps it does make sense to make the sampling function also part of the object, since it's tied to the object's RNG. Thoughts?

> Also, what do we do with _trunc_norm? That distribution isn't listed in numpy RandomState.
edit: here's a way maybe to deal with truncnorm: https://stackoverflow.com/a/50033986

Nice find. Yes, we'll have to ensure all sampling functions use a passed RandomState object.

> Also from this discussion, it seems that maybe just doing np.random.seed(some_integer) would work across all the randomness currently in Scaper, if we switch random to np.random (I think the only special change needed to make this work is random.normalvariate -> np.random.normal in SUPPORTED_DIST).

Setting `np.random.seed` would indeed work across all randomness in scaper (assuming we get rid of  python's `random` as you mention), but that ties scaper's randomness with the entire script. Imagine you have a script, then you add something non-scaper-related to script that uses `np.random` (e.g. `time.sleep(np.random.rand())`) - this will change the output of the script.

I like what sklearn does - since all randomness is determined by np.random, you *can* rely on just setting `np.random.seed` at the top of your script. However, it also supports providing objects with `RandomState` such that their random behavior is independent of any global setting. My preference would be to follow this paradigm.

**Comment by pseeth** [comment]
I think moving the distribution sampling function into the Scaper class is the cleanest way to do it - would the sampling function ever be needed outside of the Scaper class? If it was, we can just make it a public member of the class. So I propose moving the function into the class, and swapping python's random out for numpy's random. I think SUPPORTED_DIST will become a bit unwieldy if it stays as a dictionary. Thoughts on how it could change? Maybe turn it into a list and move the actual logic for each distribution to a separate function that gets called? Once that's ironed out, I can take a crack at this!

**Comment by justinsalamon** [comment]
> would the sampling function ever be needed outside of the Scaper class?

Not right now, but I can foresee adding high-level generators (methods that allow you to easily achieve something like "I need 100 soundscapes with a duration between 10-15 seconds and between 2-6 events") - these generators would live outside the scope of the `Scaper` class, and would require the sampling functions to e.g. choose how many events to add to the specification of a `Scaper` object.

As such, it might be cleaner to keep `_get_value_from_dist` global, but add an optional `random_state` kwarg following the same sklearn API (i.e. it can take an int, a RandomState object or None). Then a specific object would pass this function its `self.random_state` field when calling it. What do you think?

> I think SUPPORTED_DIST will become a bit unwieldy if it stays as a dictionary.

Why?

> Thoughts on how it could change?

One option would be to use [functools.partial](https://docs.python.org/2/library/functools.html#functools.partial) so that

```
def _get_value_from_dist(dist_tuple):
    return SUPPORTED_DIST[dist_tuple[0]](*dist_tuple[1:])
```

Turns into:

```
def _get_value_from_dist(dist_tuple, random_state=None):
   # TODO: checks on random_state
   return functools.partial(SUPPORTED_DIST[dist_tuple[0]], random_state=random_state)(*dist_tuple[1:])
```

(this assumes all the functions in the `SUPPORT_DIST` dictionary accept a `random_state` kwarg.

**Comment by justinsalamon** [comment]
@pseeth just checking in to see if you've had a chance to give this any further thought?

**Comment by justinsalamon** [comment]
> Hey, sorry, was traveling for ISMIR and now the ICASSP deadline is on me.

No worries, tis a busy time of year for us audio folk :)

> Is it okay if I do a bunch of Scaper stuff in November after the deadline?

Of course, whenever you have time.

> If you get to this before me, that's alright.

I'll let you know if I do!

> I have the source separation fork as well that I need to merge soon.

Yeah, that's going to be an awesome addition, looking forward to it.

**Comment by justinsalamon** [comment]
p.s - @pseeth quick reminder that I'd like to have a discussion and reach consensus regarding the API design and code updates involved in adding the source sep functionality. As such, the sooner you can open a PR the better, even if you don't work on it for a while (even if there was no code yet!), because it will give us time to discuss and throw ideas around. Opening the PR should be the beginning of the process, not the end :)

**Comment by justinsalamon** [comment]
Added via #54
